src/paragraphcommand.js

```diff
@@ -69,10 +72,28 @@ export default class ParagraphCommand extends Command {
 	 * @private
 	 */
 	_updateValue() {
-		const block = this.editor.document.selection.getSelectedBlocks().next().value;
+		const block = this._getSelectedBlock();
+
+		this.value = !!block && block.is( 'paragraph' );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	_checkEnabled() {
+		const block = this._getSelectedBlock();
+		const schema = this.editor.document.schema;
 
-		if ( block ) {
-			this.value = block.is( 'paragraph' );
-		}
+		return !!block && schema.check( { name: 'paragraph', inside: block.parent.name } ) && !schema.objects.has( block.name );
```

Comment:
Use with position, like this:

```js
		const isMQAllowed = schema.check( {
			name: 'messageQuote',
			inside: Position.createBefore( firstBlock )
		} );
```

Then, the schema will have entire context to check, not only the name.